File: 4.8-bc/py/stage-3-hexagonal/app/catalog/infra/postgres/product_repo.py
"""Postgres adapter for Catalog Repository."""
from __future__ import annotations

from sqlalchemy import text
from sqlalchemy.ext.asyncio import async_sessionmaker

from app.catalog.domain.product import Product


class ProductRepo:
    def __init__(self, session_maker: async_sessionmaker) -> None:
        self._session_maker = session_maker

    async def list(self) -> list[Product]:
        async with self._session_maker() as session:
            rows = (
                await session.execute(
                    text(
                        "SELECT id, name, price_cents, category_id "
                        "FROM catalog_products ORDER BY name"
                    )
                )
            ).all()
        return [
            Product(
                id=r.id,
                name=r.name,
                price_cents=r.price_cents,
                category_id=r.category_id,
            )
            for r in rows
        ]

File: 4.8-bc/ts/stage-3-hexagonal/src/catalog/infra/postgres/productRepo.ts
import type { DbPool } from "../../../shared/db.js";
import type { Product } from "../../domain/product.js";
import type { Repository } from "../../domain/repository.js";

interface ProductRow {
  id: string;
  name: string;
  price_cents: string;
  category_id: string;
}

export class PgProductRepo implements Repository {
  constructor(private readonly db: DbPool) {}

  async list(): Promise<Product[]> {
    const res = await this.db.query<ProductRow>(
      `SELECT id, name, price_cents, category_id FROM catalog_products ORDER BY name`,
    );
    return res.rows.map((r) => ({
      id: r.id,
      name: r.name,
      priceCents: Number(r.price_cents),
      categoryId: r.category_id,
    }));
  }
}
